runtime/caml/stack.h

```diff
@@ -85,21 +85,27 @@ struct caml_context {
 
 /* Structure of frame descriptors */
 
+
 typedef struct {
-  uintnat retaddr;
+  /* Frame descriptors for allocations combined with Comballoc may have
+     multiple separate debug info entries, used to separately track the
+     allocations with Gc.Memprof. These are stored as additional
+     debug_info_offset fields *before* the frame_descr, so that the i'th
+     allocation's debug info is at (&d->debug_info_offset)[-i].
+     All but the last are tagged with a 1 in the lowest bit. */
+  uint32_t debug_info_offset;
```

Comment:
Why include the last one in the struct and reorder the retaddr field? I would probably have expected all the debug_info parts to still be outside of the struct.

Reply:
Mostly, alignment issues. On 64-bit machines, the structure needs to be 8-byte aligned, so if the debuginfo was before it it would need to start at position 4 mod 8. There's no good way to portably set that up in assembly, so we end up aligning to 8 and putting 4 bytes of padding in if there are an odd number of debuginfos. Since the common case is 1 debuginfo, this grows most frame descriptors by 4 bytes.

Instead, the first one goes in the frame_descr, so we only need the 4-byte pad in even-length comballoc cases.

(This also makes the C pointer twiddling a bit nicer, as you can take the address of the first field and subtract to get the rest)